### PR TITLE
Fixed Unsigned casting error #12616

### LINKED ISSUE #12616: cast from string to uint cannot handle max(uint)
``` chapel
var u:uint = 0xffffffffffffffff;

writef("%xu\n", u);

var v = "0xffffffffffffffff":uint;
writef("%xu\n", v);
```

```
ffffffffffffffff
uncaught IllegalArgumentError: bad cast from string '0xffffffffffffffff' to uint(64)
  scast.chpl:5: thrown here
  scast.chpl:5: uncaught here
```

### PR DESCRIPTION
**Pull request by himbhadani**
This PR will Fix #12616.
Issue was in castchpl.c library in which `illegalFirstUnsChar` is called to check the invalid first char for unsigned int. but in function body, except numeric and '+' character, for all character function was returning true which was causing the casting error.  

- [x] Added a new test case test/functions/casts/stringToUint.chpl

- [x] Testcase Passed

- [x] Considered First review of @mppf

### CODE REVIEW

File: runtime/src/chplcast.c [hunk 1]
```diff
@@ -51,7 +51,7 @@ static int scanningNCounts(void) {
 }
 
 static int illegalFirstUnsChar(char c) {
-  return ((c < '0' || c > '9') && (c != '+'));
+  return ((c < '0' || c > '9') && (c != '+') && (c < 'a' || c > 'f') &&(c < 'A' || c > 'F'));
```
Comment (mppf) [inline runtime/src/chplcast.c]: I think this should just `return c == '-'`. I think trying to check against the other values is making it unnecessarily complicated. After all, we don't seem to need this level of checking in the signed case.

File: test/functions/casts/stringToUint.chpl [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+var decL = 999;
+var decS = "999": uint;
+
+var hexL = 0xFFF;
+var hexS = "0xFFF": uint;
+
+assert(binL == binS);
+assert(octL == octS);
+assert(decL == decS);
+assert(hexL == hexS);
```
Comment (mppf) [inline test/functions/casts/stringToUint.chpl]: We like to include a newline at the end of the file.

Comment (mppf) [reply test/functions/casts/stringToUint.chpl]: Please also add the example from issue #12616 to this test.

File: runtime/src/chplcast.c (hunk 1 above)
Comment (himbhadani) [reply runtime/src/chplcast.c]: Okay, I'm changing it to return  `c ==  '-'`

File: test/functions/casts/stringToUint.chpl (hunk 2 above)
Comment (himbhadani) [reply test/functions/casts/stringToUint.chpl]: okay, will add newline at the end of file

> We like to include a newline at the end of the file.
Example Written in #12616 and the example which I've included, both are almost similar. if my test will pass then example in #12616 will also pass. Should i still add the one which is in #12616 ? 
> Please also add the example from issue #12616 to this test.

**Comment by himbhadani** [comment]
Okay, I'll edit the PR description.

> You'll need to create something in the PR message. After you address my feedback, two more things will need to happen

I've already sent it on Sat, Mar 16, 4:41 PM(IST) with subject "Himanshu Bhadani's Chapel Open Source Project Individual Contributor License Agreement"
>     1. We'll need a contributor agreement -- see https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements

I had tried to run the complete test suite, but it was taking too much of time. but no problem I'll start it again. After test suite completion, how will i have to show the result?
>     2. Somebody will have to run the full test suite with the change

**Comment by mppf** [comment]
@himbhadani - we don't expect you to run the full test suite (but it's great to do if you are able). In this case I would recommend to use `start_test` to run your new test as well as test/release/examples/primers and test/types/string/cast.

When the PR is otherwise ready, I can run the full test suite using a compute cluster (so it does not take so long).

**Comment by mppf** [comment]
> how will i have to show the result?

It's good to indicate what you testing in the PR message. I like to use checkboxes, which you can do with e.g. `- [ ] new test passes`

**Comment by himbhadani** [comment]
@mppf please review.

**Comment by mppf** [comment]
This PR has passed full local testing.

**Comment by mppf** [comment]
- [ ] verify contributor agreement received

**Pull request merged**